File: Sorting/Bubble Sort/bubble_sort.py
# 
# Created by @PythonicVarun on 11-11-2024.
# 

from typing import List


def bubble_sort(arr: List, reverse: bool = False) -> None:
    """
    Sorts an array using the bubble sort algorithm with an option for reverse order.

    This function sorts an array of integers in ascending or descending order
    based on the value of the `reverse` parameter. If `reverse` is False, it sorts
    in ascending order. If `reverse` is True, it sorts in descending order.

    Time Complexity:
        - Worst and Average Case: O(n^2) when the array is unsorted.
        - Best Case: O(n) when the array is already sorted (due to the early exit optimization).

    Space Complexity: O(1), as the sorting is done in-place.

    Parameters:
        arr (list): The array to be sorted.
        reverse (bool): If False, sorts in ascending order; if True, sorts in descending order.

    Returns:
        None: The function sorts the array in place.
    """
    n = len(arr)
    for i in range(n):
        flag = False
        for j in range(n - i - 1):
            if (not reverse and arr[j] > arr[j + 1]) or (reverse and arr[j] < arr[j + 1]):
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                flag = True
        if flag is True:
            return  # If no elements were swapped, the array is already sorted

def main():
    """
    Main function to test the bubble sort algorithm with reverse functionality.

    This function initializes an array of integers, sorts it using bubble sort
    in both ascending and descending order, and prints the sorted arrays.

    Returns:
        None
    """
    arr = [64, 25, 12, 99, 22, 11]
    
    print("Original array:", arr)
    
    # Sort in ascending order
    bubble_sort(arr, reverse=False)
    print("Sorted array (ascending):", arr)
    
    # Sort in descending order
    bubble_sort(arr, reverse=True)
    print("Sorted array (descending):", arr)

if __name__ == "__main__":
    main()

File: Sorting/Bubble Sort/bubble_sort.c
//
// Created by @PythonicVarun on 11-11-2024.
//

#include <stdio.h>

/**
 * @brief Swaps the values of two integers.
 *
 * This function takes pointers to two integers and swaps their values.
 *
 * @param a Pointer to the first integer.
 * @param b Pointer to the second integer.
 */
void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * @brief Sorts an array using the bubble sort algorithm with an option for reverse order.
 *
 * This function sorts an array of integers in ascending or descending order using the
 * bubble sort algorithm. If the `reverse` parameter is 0, it sorts in ascending order.
 * If `reverse` is 1, it sorts in descending order.
 *
 * @note Time Complexity:
 *     - Worst and Average Case: O(n^2) when the array is unsorted.
 *     - Best Case: O(n) when the array is already sorted (due to early exit optimization).
 *
 * @note Space Complexity: O(1), since the sorting is done in-place.
 * 
 * @param arr[] The array to be sorted.
 * @param n The number of elements in the array.
 * @param reverse If 0, sorts in ascending order; if 1, sorts in descending order.
 */
void bubbleSort(int arr[], int n, int reverse) {
    for (int i = 0; i < n; i++) {
        int flag = 0;
        for (int j = 0; j < n - i - 1; j++) {
            if ((reverse == 0 && arr[j] > arr[j + 1]) || (reverse == 1 && arr[j] < arr[j + 1])) {
                flag = 1;
                swap(&arr[j], &arr[j + 1]);
            }
        }
        if (flag == 0) return;  // If no elements were swapped, the array is sorted
    }
}

/**
 * @brief Main function to test the bubble sort algorithm with reverse functionality.
 *
 * This function initializes an array of integers, sorts it using
 * bubble sort in both ascending and descending order, and prints the sorted arrays.
 *
 * @return int Program exit status.
 */
int main() {
    int arr[] = {64, 25, 12, 22, 11};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Original array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Sort in ascending order
    bubbleSort(arr, n, 0);
    printf("Sorted array (ascending): ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Sort in descending order
    bubbleSort(arr, n, 1);
    printf("Sorted array (descending): ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}
